src/core/server/base.js

```diff
@@ -302,10 +303,34 @@ function baseServer(routes, createStore, { appInstanceName = appName } = {}) {
             throw errorPage.error;
           }
 
-          return hydrateOnClient({
-            props: { component: InitialComponent },
-            pageProps,
-            res,
+          const props = { component: InitialComponent };
+
+          // TODO: Remove the try/catch block once all apps are using
+          // redux-saga.
+          let rootSaga;
+          try {
+            // eslint-disable-next-line global-require, import/no-dynamic-require
+            rootSaga = require(`${appName}/sagas`).default;
+          } catch (err) {
+            log.warn(`sagas not found for ${appName}`, err);
+          }
+
+          if (!rootSaga) {
+            return hydrateOnClient({ props, pageProps, res });
+          }
+
+          const sagas = store.runSaga(rootSaga);
+          // We need to render once because it will force components
+          // with sagas to call the sagas and load their data.
+          ReactDOM.renderToString(<ServerHtml {...pageProps} {...props} />);
```

Comment:
If that happens we'll be rendering HTML that doesn't have any data loaded from sagas. If you see the comments, we actually have to render twice when doing sagas on the server.

The first render will get any empty components to dispatch an action to fetch data with sagas. Once they're finished we have all that data in the store and THEN we can render and actually use the data.

Because we always need to render in `hydrateOnClient` (and it's called in a few places as the "final page load" function) we can just render after loading the sagas once to get them to dispatch all their actions.